wicket/src/state/update.rs

```diff
@@ -103,6 +127,162 @@ impl RackUpdateState {
             }
         }
     }
+
+    /// Scan through `log` and update the components status given by `id`
+    pub fn update_items(&mut self, id: &ComponentId, log: &UpdateLog) {
+        let items = self.items.get_mut(id).unwrap();
+        if log.events.is_empty() {
+            // Reset all items to default
+            for (_, state) in items.iter_mut() {
+                *state = UpdateState::Waiting;
+            }
+        }
+
+        // Mark artifacts as either 'succeeded' or `failed' by looking in
+        // the event log.
+        for event in &log.events {
+            match &event.kind {
+                UpdateEventKind::Normal(normal) => match normal {
+                    UpdateNormalEventKind::SpResetComplete => {
+                        let known = Some(id.sp_known_artifact_kind());
+                        update_artifact_state(
+                            items,
+                            known,
+                            UpdateState::Updated,
+                        );
+                    }
+                    UpdateNormalEventKind::ArtifactUpdateComplete {
+                        artifact,
+                    } => {
+                        // We specifically don't want to mark the Host
+                        // complete, if the trampoline has completed, so don't
+                        // use `artifact_to_known_artifact_kind`
+                        let known = artifact.kind.parse().ok();
+                        update_artifact_state(
+                            items,
+                            known,
+                            UpdateState::Updated,
+                        );
+                    }
+                    UpdateNormalEventKind::InstallinatorEvent(
+                        _completion_event_kind,
+                    ) => {
+
+                        // TODO: Do we want to update state
+                        // with intermediate events?
+                        //
+                        // This would be useful for reporting progress
+                        // percentages, but that's about it.
```

Comment:
Can we represent "downloaded" and "installed" as separate events here? That would fit nicely here then, I think

Reply:
Maybe. One problem is that this doesn't really work for the Sp and RoT which write while they are downloading. It's all streaming. I'll merge this through and we can discuss in a follow up. There's also what @jgallagher suggested which is in the status column to indicate for sleds which thing was being updated. I didn't do that just due to the current structure of the code where `UpdateState` is an enum and all states are currently valid for all entities. But that could change and may line up with this here.